internal/internal_worker.go

```diff
@@ -619,6 +622,33 @@ func (r *registry) RegisterWorkflowWithOptions(
 	}
 }
 
+func (r *registry) RegisterDynamicWorkflow(wf interface{}, options DynamicRegisterOptions) {
+	// Support direct registration of WorkflowDefinition
+	factory, ok := wf.(WorkflowDefinitionFactory)
+	if ok {
+		r.Lock()
+		defer r.Unlock()
+		r.dynamicWorkflow = factory
+		r.dynamicOptions = options
+		return
+	}
+
+	// Validate that it is a function
+	fnType := reflect.TypeOf(wf)
+	if err := validateFnFormat(fnType, true, true); err != nil {
+		panic(err)
+	}
+
+	if r.dynamicWorkflow != nil {
```

Comment:
Not concurrency safe. There's nothing blocking in this function, I'd just move the lock/unlock statements to be the first two statements in the function, no need to try to narrow it